diff -Naur snappy/snappy.cc snappy_patched/snappy.cc
--- snappy/snappy.cc	2022-04-12 20:44:55.000000000 -0400
+++ snappy_patched/snappy.cc	2022-04-12 20:47:05.000000000 -0400
@@ -1014,7 +1014,7 @@
 }
 
 SNAPPY_ATTRIBUTE_ALWAYS_INLINE
-size_t AdvanceToNextTag(const uint8_t** ip_p, size_t* tag) {
+inline size_t AdvanceToNextTag(const uint8_t** ip_p, size_t* tag) {
   const uint8_t*& ip = *ip_p;
   // This section is crucial for the throughput of the decompression loop.
   // The latency of an iteration is fundamentally constrained by the
